cli/src/maintenance.rs

```diff
@@ -576,32 +624,49 @@ impl SolidoState {
+            self.solido.validators.entries.iter(),
+            self.validator_stake_accounts.iter(),
+            self.validator_unstake_accounts.iter()
+        ) {
+            let current_stake_balance = stake_accounts
                 .iter()
                 .map(|(_addr, detail)| detail.balance.total())
                 .sum::<token::Result<Lamports>>()
                 .expect("If this overflows, there would be more than u64::MAX staked.");
 
-            if current_balance > validator.entry.stake_accounts_balance {
-                let expected_difference = (current_balance
-                    - validator.entry.stake_accounts_balance)
-                    .expect("Does not overflow because current > entry.balance.");
-                // If the expected difference is less than some defined amount
-                // of Lamports, we don't bother withdrawing. We try to do this
-                // so we don't pay more for fees than the amount that we'll
-                // withdraw.
-                if expected_difference < SolidoState::MINIMUM_WITHDRAW_AMOUNT {
-                    continue;
+            let expected_difference_stake =
+                if current_stake_balance > validator.entry.stake_accounts_balance {
+                    (current_stake_balance - validator.entry.stake_accounts_balance)
+                        .expect("Does not overflow because current > entry.balance.")
+                } else {
+                    Lamports(0)
+                };
+
+            let mut removed_unstake = Lamports(0);
+
+            for (_addr, unstake_account) in unstake_accounts.iter() {
+                if unstake_account.balance.inactive != unstake_account.balance.total() {
+                    break;
                 }
+                removed_unstake = (removed_unstake + unstake_account.balance.total())
+                    .expect("Summing unstake accounts should not overflow.");
+            }
+
+            // If the expected difference is less than some defined amount
+            // of Lamports, we don't bother withdrawing. We try to do this
+            // so we don't pay more for fees than the amount that we'll
+            // withdraw. Or if we have stake to remove from unstake accounts.
+            if expected_difference_stake > SolidoState::MINIMUM_WITHDRAW_AMOUNT
+                || removed_unstake > Lamports(0)
+            {
                 // The balance of this validator is not up to date, try to update it.
-                let stake_account_addrs = stake_accounts.iter().map(|(addr, _)| *addr).collect();
+                let mut stake_account_addrs: Vec<Pubkey> =
+                    stake_accounts.iter().map(|(addr, _)| *addr).collect();
+                // Try to also withdraw from unstake accounts
+                let mut unstake_account_addrs: Vec<Pubkey> =
+                    unstake_accounts.iter().map(|(addr, _)| *addr).collect();
+                stake_account_addrs.append(&mut unstake_account_addrs);
```

Comment:
It’s possible to collect directly into the vec without an intermediate vec:

```rust
let mut stake_account_addrs = Vec::new();
stake_account_addrs.extend(stake_accounts.iter().map(|(addr, _)| *addr));
stake_account_addrs.extend(unstake_accounts.iter().map(|(addr, _)| *addr));
```

Reply:
Much better, thanks!